### PR TITLE
Lowering softmax

### PR DESCRIPTION
**Pull request by tungld**
This patch is to lower ONNXSoftmaxOp, conforming to its description here: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Softmax

### CODE REVIEW

**Comment by tjingrant** [comment]
Sorry for the wait, I'm trying to enable end-to-end value-based tests on Buildbot.
I had it working on our internal ones, but haven't moved everything here.

**Comment by tungld** [comment]
I looked at the internal ones and saw that you commented some failed tests about Min, Max, Sum. I guess we can fix them by #16.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 1]
```diff
@@ -824,6 +824,225 @@ struct ONNXElementwiseVariadicOpLowering : public ConversionPattern {
...
+struct ONNXSoftmaxOpLowering : public ConversionPattern {
+  ONNXSoftmaxOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXSoftmaxOp::getOperationName(), 1, ctx) {}
+  PatternMatchResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const final {
+    // softmax(x) = let max_x = max(x) in
+    //                let exp_x = exp(x - max_x) in
+    //                  let sum = sum(exp_x) in
+    //                    exp_x / sum
```
Comment (KeDengMS) [inline src/pass/lower_frontend_to_krnl.cpp]: Softmax contains two Reduce ops: ReduceMax and ReduceSum. Would it better to have lowings of those ops implemented and let Softmax reuse?

Comment (tungld) [reply src/pass/lower_frontend_to_krnl.cpp]: @KeDengMS Yes, I agree. This patch manually lowers softmax, we will replace it with a higher-level rewriting you mentioned once ReduceMax and ReduceSum are available.

File: src/pass/lower_frontend_to_krnl.cpp [hunk 2]
```diff
@@ -824,6 +824,225 @@ struct ONNXElementwiseVariadicOpLowering : public ConversionPattern {
...
+    auto tensorType = (*op->result_type_begin()).cast<RankedTensorType>();
+    int64_t rank = tensorType.getRank();
+    int64_t axis = op->getAttrOfType<IntegerAttr>("Softmax.axis").getInt();
```
Comment (tjingrant) [inline src/pass/lower_frontend_to_krnl.cpp]: I think it should be just "axis"? instead of "Softmax.axis".

Comment (tungld) [reply src/pass/lower_frontend_to_krnl.cpp]: In the current version, it is still "Softmax.axis". https://github.com/clang-ykt/ONNF/blob/0aaab0d2d2aa9580c069a64642b727e804abf4da/src/builder/frontend_dialect_transformer.cpp#L246

**Pull request merged**